storage/types.go

```diff
@@ -79,16 +80,39 @@ func NewNodeIDWithPrefix(prefix uint64, prefixLenBits, nodeIDLenBits, maxLenBits
 	return p
 }
 
+func bitLen(x int64) int {
+	r := 0
+	for x > 0 {
+		r++
+		x >>= 1
+	}
+	return r
+}
+
 // NewNodeIDForTreeCoords creates a new NodeID for a Tree node with a specified depth and
-// index
-func NewNodeIDForTreeCoords(depth int64, index int64, maxLenBits int) NodeID {
-	r := NewEmptyNodeID(maxLenBits)
-	for i := len(r.Path) - 1; index > 0; i-- {
+// index.
+// This method is used exclusively by the Log, and since the Log model grows upwards from the
+// leaves we modify the provided coords accordingly.
+//
+// index is the horizontal index into the tree at level depth, so the returned
+// NodeID will be zero padded on the right by depth places.
+func NewNodeIDForTreeCoords(depth int64, index int64, maxTreeDepth int) (NodeID, error) {
+	bl := bitLen(index)
+	if depth < 0 || bl > int(maxTreeDepth-int(depth)) {
```

Comment:
Have I understood correctly:   Does depth start at 0 from the root of the tree, and index start from 0 at each level of the tree, and go along to 2^depth -1?  If so, I'm not convinced by the bl > int(maxTreeDepth - int(depth)) condition here - suppose the tree had 2^64 leaves, and you're trying to get the NodeID for the very last leaf - i.e. at index 2^64-1 on at depth 64 (double check the numbers because they could be wrong :P)  This would give bl = 64 which is > 64 - 64 = 0... would that not be a valid depth/index combo?

New to this code so please bear with me if I'm not understanding correctly!